### PR TITLE
Charts date ranges

### PR DESCRIPTION
**Pull request by SebaKas21**
Hi @dawiddworak88 

I know how to transfer data from fe to the controller in the web. Thanks, but the problem is elsewhere. 

Exactly the problem is that the ApiClientService and the GetAsync method converts the date to a different format.
Date transferred to Repository: 2023.02.27 00:48
Date transferred from Repository to Api: 02.27.2023 00:48

URL going to analytics-api: http://analytics-api/api/v1/salesanalytics/countries?FromDate=26.11.2022+23%3a40%3a10&ToDate=26.02.2023+23%3a40%3a10

Response is: Bad Request

EDIT:

As I checked in Postman, the first date format works, but nothing works with the second one either.

Response in Seller.Web

obraz

### CODE REVIEW

**Review by dawiddworak88 — CHANGES_REQUESTED** [review]
Hi @SebaKas21, I adjusted the date time handling in ApiClient Service. I'd rather not create a PR yet because it still needs a lot of work and code clean up: 

image

**Comment by SebaKas21** [comment]
Hi @dawiddworak88  

I know, but I wanted to show you the problem that was occurring.

Thank you very much for your help ;)

File: be/src/Project/Services/Analytics/Analytics.Api/Validators/GetAnnualSalesModelValidator.cs [hunk 1]
```diff
@@ -9,6 +10,22 @@ public class GetAnnualSalesModelValidator : BaseServiceModelValidator<GetAnnualS
...
+            this.RuleFor(x => x).Must(y =>
+            {
+                if (y.FromDate > y.ToDate)
+                {
+                    return false;
+                }
+
+                var currentDate = DateTime.UtcNow;
+
+                if ((y.FromDate > currentDate) && (y.ToDate > currentDate))
```
Comment (dawiddworak88) [inline be/src/Project/Services/Analytics/Analytics.Api/Validators/GetAnnualSalesModelValidator.cs]: **suggestion:** delete '&& (y.ToDate > currentDate)'

File: be/src/Project/Services/Analytics/Analytics.Api/Validators/GetAnnualSalesModelValidator.cs [hunk 2]
```diff
@@ -9,6 +10,22 @@ public class GetAnnualSalesModelValidator : BaseServiceModelValidator<GetAnnualS
         public GetAnnualSalesModelValidator()
         {
             this.RuleFor(x => x.OrganisationId).NotEmpty().NotNull();
+            this.RuleFor(x => x).Must(y =>
+            {
+                if (y.FromDate > y.ToDate)
```
Comment (dawiddworak88) [inline be/src/Project/Services/Analytics/Analytics.Api/Validators/GetAnnualSalesModelValidator.cs]: **suggestion:** put it all in once condition

File: be/src/Project/Services/Analytics/Analytics.Api/Validators/GetDailySalesModelValidator.cs [hunk 3]
```diff
@@ -9,6 +10,22 @@ public class GetDailySalesModelValidator : BaseServiceModelValidator<GetDailySal
         public GetDailySalesModelValidator()
         {
             this.RuleFor(x => x.OrganisationId).NotEmpty().NotNull();
+            this.RuleFor(x => x).Must(y =>
```
Comment (dawiddworak88) [inline be/src/Project/Services/Analytics/Analytics.Api/Validators/GetDailySalesModelValidator.cs]: **suggestion:** add inheritance not to repeat code

File: be/src/Project/Services/Analytics/Analytics.Api/Validators/GetAnnualSalesModelValidator.cs (hunk 2 above)
Comment (dawiddworak88) [inline be/src/Project/Services/Analytics/Analytics.Api/Validators/GetAnnualSalesModelValidator.cs]: **suggestion:** add check if fromDate or toDate is not null, empty, zeros or DateTime.MinDate

File: be/src/Project/Services/Analytics/Analytics.Api/v1/Controllers/SalesAnalyticsController.cs [hunk 4]
```diff
@@ -75,46 +81,58 @@ public IActionResult GetAnnualSales()
...
         /// Get daily sales
         /// </summary>
+        /// <param name="fromDate">From date.</param>
+        /// <param name="toDate">To date.</param>
         /// <returns>Daily sales.</returns>
         [HttpGet("daily"), MapToApiVersion("1.0")]
         [ProducesResponseType((int)HttpStatusCode.OK)]
         [ProducesResponseType((int)HttpStatusCode.UnprocessableEntity)]
-        public IActionResult GetDailySales()
+        public IActionResult GetDailySales(DateTime fromDate, DateTime toDate)
```
Comment (dawiddworak88) [inline be/src/Project/Services/Analytics/Analytics.Api/v1/Controllers/SalesAnalyticsController.cs]: **suggestion:** make them nullable and set a default value of toDate to UtcNow If null

File: be/src/Project/Web/Seller/Seller.Web/Areas/Dashboard/ApiControllers/DailySalesAnalyticsApiController.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            {
+                var chartLabels = new List<string>();
+
+                foreach (var dailySalesItem in dailySales)
+                {
+                    var dayName = CultureInfo.CurrentUICulture.DateTimeFormat.GetDayName(((DayOfWeek)dailySalesItem.DayOfWeek));
+
+                    var monthNumber = dailySalesItem.Month.ToString();
+
+                    if (dailySalesItem.Month < 10)
```
Comment (dawiddworak88) [inline be/src/Project/Web/Seller/Seller.Web/Areas/Dashboard/ApiControllers/DailySalesAnalyticsApiController.cs]: **suggestion:** move to constants and add descriptive name

File: be/src/Project/Web/Seller/Seller.Web/Areas/Dashboard/ModelBuilders/DailySalesAnalyticsModelBuilder.cs [hunk 6]
```diff
@@ -17,25 +18,40 @@ namespace Seller.Web.Areas.Dashboard.ModelBuilders
...
             _dashboardResources = dashboardResources;
             _salesAnalyticsRepository = salesAnalyticsRepository;
+            _linkGenerator = linkGenerator;
+            _globalResources = globalResources;
         }
 
         public async Task<DailySalesAnalyticsViewModel> BuildModelAsync(ComponentModelBase componentModel)
         {
-            var dailySales = await _salesAnalyticsRepository.GetDailySales(componentModel.Token, componentModel.Language);
+            var fromDate = DateTime.UtcNow.AddDays(-21);
```
Comment (dawiddworak88) [inline be/src/Project/Web/Seller/Seller.Web/Areas/Dashboard/ModelBuilders/DailySalesAnalyticsModelBuilder.cs]: **suggestion:** remove magic strings from code

**Comment by SebaKas21** [comment]
Hi @dawiddworak88 

You haven't published your review

**Comment by SebaKas21** [comment]
Hi @dawiddworak88 

I think you can take a look at it now :)

**Review by dawiddworak88 — APPROVED** [review]

**Pull request merged**